src/subscriber.ts

```diff
@@ -170,52 +183,113 @@ export class Subscriber {
 
-     this.validators.forEach(async account => {
-        if( isHeartbeatExpected && ! await this._hasValidatorAuthoredBlocks(account,sessionIndex) && ! await this._hasValidatorSentHeartbeat(account,sessionIndex) ){
-          this.logger.info(`Target ${account.name} has either not authored any block or sent any heartbeat yet`);
-          this.promClient.setStateValidatorOfflineSessionReports(account.name)
+        const validatorActiveSetIndex = this._getValidatorActiveSetIndex(account)
+        if ( validatorActiveSetIndex < 0 ) {
+          this._handleValidatorOutOfActiveSetTrue(account)
+          return 
         }
         else{
-            this.promClient.resetStateValidatorOfflineSessionReports(account.name)
+          this._handleValidatorOutOfActiveSetFalse(account)
         }
-      })
+
+        if(isHeartbeatExpected) {
+          if ( await this._hasValidatorProvedOnline(account,validatorActiveSetIndex,sessionIndex) ) {
+            this.promClient.resetStatusValidatorOffline(account.name);
+          }
+          else {
+            this.logger.info(`Target ${account.name} has either not authored any block or sent any heartbeat yet`);
+            this.promClient.setStatusValidatorOffline(account.name);
+          }
+        }
+        else if ( this.promClient.isValidatorStatusOffline(account.name) ) {
+          if ( await this._hasValidatorProvedOnline(account,validatorActiveSetIndex, sessionIndex) ){
+            this.promClient.resetStatusValidatorOffline(account.name);
+          }
+        }
+
+      }) 
     }
 
-    private async _subscribeOffline(): Promise<void> {
-        this.validators.forEach((account) => {
-            // always increase metric even the first time, so that we initialize the time serie
-            // https://github.com/prometheus/prometheus/issues/1673
-            this.promClient.resetTotalValidatorOfflineReports(account.name);
-        });
+    private _handleValidatorOutOfActiveSetTrue(account: Subscribable): void{
+      this.logger.debug(`Target ${account.name} is not presente in the current validation active set`);
+      this.promClient.resetStatusValidatorOffline(account.name);
+      this.promClient.setStatusValidatorOutOfActiveSet(account.name);
+    }
+
+    private _handleValidatorOutOfActiveSetFalse(account: Subscribable): void{
+      this.promClient.resetStatusValidatorOutOfActiveSet(account.name);
+    }
+
+    private async _hasValidatorProvedOnline(account: Subscribable, validatorIndex: number, sessionIndex: SessionIndex): Promise<boolean> {
+      return await this._hasValidatorAuthoredBlocks(account,sessionIndex) || await this._hasValidatorSentHeartbeat(validatorIndex,sessionIndex)
+    }
+
+    private async _subscribeEvents(): Promise<void> {
+        this.subscribe.offline && this._initTotalValidatorOffline();
 
         this.api.query.system.events((events) => {
-            events.forEach((record) => {
+            events.forEach(async (record) => {
```

Comment:
I'll double check. What I want to avoid is to use the asyncForEach utils function + the await on each execution: there is no need to enforce a serialization in this context. Wdyt ?

Reply:
Makes sense, we don't care about the order in which they are executed :+1: